src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1884,11 +1884,21 @@ internal async Task CloseOngoingOperations(bool async)
         }
     }
 
-    // TODO in theory this should be async-optional, but the only I/O done here is the Terminate Flush, which is
-    // very unlikely to block (plus locking would need to be worked out)
     internal void Close()
     {
-        lock (this)
```

Comment:
This particular lock is actually very important. It helps at least with:
1. Concurrent close
2. Concurrent keepalive
3. Concurrent break

I'm not sure whether changing everywhere to use `SemaphoreSlim` is a good idea, would have to think about that.

Reply:
Yeah, still no idea what to do here. @roji maybe you have something on your mind?